src/storage-client/src/client.rs

```diff
@@ -354,6 +354,65 @@ impl PackableStats for SinkStatisticsUpdate {
     }
 }
 
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Status {
+    Starting,
+    Running,
+    Paused,
+    Stalled,
+    Ceased,
+    Dropped,
+}
+
+impl Status {
+    pub fn to_str(&self) -> &'static str {
+        match self {
+            Status::Starting => "starting",
+            Status::Running => "running",
+            Status::Paused => "paused",
+            Status::Stalled => "stalled",
+            Status::Ceased => "ceased",
+            Status::Dropped => "dropped",
+        }
+    }
+
+    /// Determines if a new status should be produced in context of a previous
+    /// status.
+    pub fn superseded_by(self, new: Status) -> bool {
+        match (self, new) {
+            (Status::Dropped, _) => false,
+            (_, Status::Dropped) => true,
+            (Status::Ceased, _) => false,
+            (_, Status::Ceased) => true,
+            // TODO(guswynn): Ideally only `failed` sources should not be marked as paused.
+            // Additionally, dropping a replica and then restarting environmentd will
+            // fail this check. This will all be resolved in:
+            // https://github.com/MaterializeInc/materialize/pull/23013
```

Comment:
I think if we return `true` in this case, this is resolved!

Reply:
sgtm